pages/common/options-loader.ts

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: Apache-2.0
+
+import { useRef, useState } from 'react';
+import { AutosuggestProps } from '~components/autosuggest';
+
+interface PendingRequest {
+  resolve: (value: APIResponse) => void;
+  reject: (reason: any) => void;
+  params: {
+    filteringText: string;
+    pageNumber: number;
+  };
+}
+
+interface APIResponse {
+  items: NonNullable<unknown[]>;
+  hasNextPage: boolean;
+}
+
+// Pending requests are used to override the result programmatically in integration tests:
+// > window.__pendingRequests.shift().resolve({ items, hasNextPage });
+interface ExtendedWindow extends Window {
+  __pendingRequests: Array<PendingRequest>;
+}
+declare const window: ExtendedWindow;
+const pendingRequests: Array<PendingRequest> = (window.__pendingRequests = []);
+
+export interface OptionsLoaderProps {
+  pageSize?: number;
+  timeout?: number;
+  randomErrors?: boolean;
+}
+
+export interface OptionsLoaderState<Item> {
+  data: readonly Item[];
+  status: AutosuggestProps.StatusType;
+}
+
+export interface FetchItemsProps<Item> {
+  firstPage: boolean;
+  filteringText: string;
+  sourceItems?: readonly Item[];
+}
+
+export function useOptionsLoader<Item>({ pageSize = 25, timeout = 1000, randomErrors = false }: OptionsLoaderProps) {
+  const [items, setItems] = useState(new Array<Item>());
+  const [status, setStatus] = useState<'pending' | 'loading' | 'finished' | 'error'>('pending');
+  const [filteringText, setFilteringText] = useState('');
+  const requestsRef = useRef(new Array<FakeRequest<APIResponse>>());
+
+  function cancelRequests() {
+    requestsRef.current.forEach(rq => (rq.cancelled = true));
+    requestsRef.current = [];
+  }
+
+  function hasPendingRequest() {
+    return requestsRef.current.some(rq => !rq.resolved && !rq.rejected);
+  }
+
+  function getNextPageNumber() {
+    const resolvedRequests = requestsRef.current.filter(rq => rq.resolved);
+    const lastResolved = resolvedRequests[resolvedRequests.length - 1];
+    return !lastResolved ? 0 : lastResolved.pageNumber + 1;
+  }
+
+  function fetchSourceItems(pageNumber: number, sourceItems: readonly Item[]) {
+    return new FakeRequest<APIResponse>(pageNumber, (resolve, reject) =>
+      setTimeout(() => {
+        if (randomErrors && Math.random() < 0.3) {
+          reject();
+        } else {
+          const nextItems = sourceItems.slice(pageNumber * pageSize, (pageNumber + 1) * pageSize);
+          resolve({ items: nextItems, hasNextPage: items.length + nextItems.length < sourceItems.length });
+        }
+      }, timeout)
+    );
+  }
+
+  function issuePendingRequest(pageNumber: number, filteringText: string) {
+    const params = { filteringText, pageNumber };
+    return new FakeRequest<APIResponse>(pageNumber, (resolve, reject) =>
+      pendingRequests.push({ resolve, reject, params })
+    );
+  }
+
+  function fetchItems({ sourceItems, firstPage, filteringText: nextFilteringText }: FetchItemsProps<Item>) {
```

Comment:
Yes, I agree. However, passing that as a hook argument won't work because this way it will be delayed by one render cycle for consecutive loading case